src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,236 @@
+// TODO
+
+/+
+
+* It should support records with fields of any type. (The old one just supports
+  records of `<char[]>` and `<char[], char[]>`.)
+* It should provide a method to get the batched size of a record. (The old one
+  has this function internally, but it's private.)
+* Compression should be optional.
+* The maximum batch size (== the record size limit) should be soft and
+  configurable at runtime, not a compile-time constant. This way, the size limit
+  could be set in the node's config file, which could then be re-parsed if the
+  limit needed to be changed. Future batches would then use the new size limit.
+* Instead of checking that each added record fits in the remaining space in the
+  batch, check that the size of each added record is <= the maximum batch size.
+  This will lead to flexible sized batches of at most 2x the maximum size. (For
+  the advantage of this approach, see
+  https://github.com/sociomantic-tsunami/swarm/issues/158#issuecomment-339671247.)
+
++/
+
+module swarm.neo.util.Batch;
+
+import ocean.transition;
+import ocean.io.compress.Lzo;
+
+struct BatchWriter
+{
+    import ocean.core.Traits;
+
+    private void[]* buffer;
+
+    // Has two functions:
+    //   1. Size limit per tuple added.
+    //   2. Once a batch exceeds this size, it is considered full.
+    private size_t max_size;
+
+    public void initialise ( void[]* buffer, size_t max_size )
+    {
+        this.buffer = buffer;
+        this.max_size = max_size;
+    }
+
+    public size_t sizeOf ( T ... ) ( T items )
+    {
+        size_t sum;
+        foreach ( item; items )
+            sum += this.sizeOfItem(item);
+
+        return sum;
+    }
+
+    public bool fits ( T ... ) ( T items )
+    {
+        return this.sizeOf(items) <= this.max_size;
+    }
+
+    // Returns: true if added, false if too big
+    public bool add ( T ... ) ( T items, void delegate ( ) send_batch )
+    {
+        if ( !this.fits(items) )
+            return false;
+
+        foreach ( item; items )
+            this.addItem(item);
+
+        if ( this.buffer.length >= this.max_size )
+            send_batch();
+
+        return true;
+    }
+
+    public Const!(void)[] get ( )
+    {
+        return *this.buffer;
+    }
+
+    public Const!(void)[] getCompressed ( Lzo lzo, ref void[] compress_buf )
+    {
+        // Set destination to max possible length.
+        compress_buf.length =
+            lzo.maxCompressedLength(this.buffer.length) + size_t.sizeof;
+        enableStomping(compress_buf);
```

Comment:
I was told `enableStomping` is expensive. Since it needs to be called only before increasing the buffer length, you could call it only in that case. Since LZO normally compresses the data one `enableStomping` call can usually be avoided.


if LZO compresses the data so that `compress_buf.length > (compressed_len + size_t.sizeof)`, then one `enableStomping` call can be avoided.